### ISSUE TITLE
A weird bug with Falkon

### ISSUE THREAD

**Issue by trmdi** [tsujan/Kvantum#435]
It only happens with Falkon.
It doesn't happen when Falkon is in the opaque list in KV Manager.

![Screenshot](https://i.imgur.com/L1AQn3I.png)
Kvantum theme: KvMojave.  (the default config)

**Comment by tsujan** [comment]
Yes, a nasty bug in Qt 5.13. I added a workaround for it.

Please upgrade to the latest git Kvantum and tell me whether the problem is gone!

**Comment by tsujan** [comment]
BTW, for a short explanation, read this code comment: https://github.com/tsujan/Kvantum/blob/79aafa023e57e8cb10bc4bbce3c100724d775bec/Kvantum/style/polishing.cpp#L1125

**Comment by trmdi** [comment]
Oops... Updated to the latest commit: 79aafa0
![sc](https://i.imgur.com/K5Wy05Y.png)

**Comment by tsujan** [comment]
This is very strange!

I started to see the issue after upgrading to Qt 5.13.1 (no Falkon or KDE upgrade) but it has been totally disappeared since I added the workaround. Falkon is my default browser.

Here, I have Qt 5.13.1, kwin 5.16.5 and Falkon 3.1.0.

Are you sure you restarted Falkon after upgrading Kvantum? No hidden process? Sorry for the obvious question but sometimes nothing should be considered obvious ;)

**Comment by trmdi** [comment]
Yes, I had restarted my computer before posting that.

**Comment by tsujan** [comment]
What's the name of the plugin that has replaced Falkon's title-bar in your screenshots?

**Comment by trmdi** [comment]
It does not replace Fankon's titlebar, those are Latte dock and Plasma widgets to show window buttons, title, appmenu... the real one is just hidden when Falkon is maximized.

**Comment by tsujan** [comment]
I want to make my system as similar to yours as possible, to see if the problem reappears. Will report back soon...

**Comment by tsujan** [comment]
I installed and activated all of them. Falkon is still OK:

![my_falkon](https://user-images.githubusercontent.com/981076/66269648-aafcbe00-e857-11e9-97d4-980726591bf3.png)

If only I could know what's the difference in your case...

**Comment by trmdi** [comment]
Let see if other Tumbleweed users meet the bug: https://www.reddit.com/r/openSUSE/comments/de3aej/do_you_tumbleweedkdekvantumfalkon_users_meet_this/ :/

**Comment by tsujan** [comment]
OK.

The strangest thing is that I did see the problem (3 weeks ago), found its cause (in Qt 5.13.1), worked around it, and it disappeared once for all. Now, your screenshots show it again, while it doesn't exist here. The question is: What's the difference?

**Comment by NickLion** [comment]
openSUSE Tumbleweed, KVantum git.20191001T052724.79aafa0, same strange behaviour with KvMojave.

**Comment by tsujan** [comment]
@NickLion Thanks!

I'm afraid I can't do anything unless I can reproduce it here. Please tell me if you have any idea what the difference may be.

Have you installed the latest git Kvantum directly from the source (`mkdir build && cd build`, then `cmake ..`, then `make`, and finally `sudo make install`)?

**Comment by wxl** [comment]
Just to throw in a little suggestion in hopes of making this reproducible, figure out the steps it needs to take to make this occur in a fresh install of Tumbleweed, e.g. in a vm.

**Comment by NickLion** [comment]
> Please tell me if you have any idea what the difference may be.

Well, openSUSE has some patches for Qt, maybe some conflict, though I haven't figure out what could make it. You've mentioned Qt bug, which needs workaround, does it exists in https://bugreports.qt.io/ ? If no, could you describe it? I'm not exactly getting meaning of workaround (I'm not familiar with QStyle interns).


> Have you installed the latest git Kvantum directly from the source (`mkdir build && cd build`, then `cmake ..`, then `make`, and finally `sudo make install`)?

No, but I'm building it without any changes and just packaging into RPM result of `make install` and docs. I do understand there might be a difference, but I've never hit any for Kvantum, its build is straightforward. I could try it later if you insist, but I doubtful it would behave differently.

**Comment by NickLion** [comment]
@trmdi As I'm not using KvMojave (or other affected theme), just tested it, could you say, is this bug new or it was some time before?

**Comment by tsujan** [comment]
@wxl I really prefer not to install 4 gigabytes of Tumbleweed on virtualbox but I might have no other choice ;)

@NickLion I don't think it's because of openSUSE's patches but I can't be sure either. [The comment I mentioned above](https://github.com/tsujan/Kvantum/blob/79aafa023e57e8cb10bc4bbce3c100724d775bec/Kvantum/style/polishing.cpp#L1125) explains the cause briefly (and also why I only rarely report bugs to Qt bug tracker).

**Comment by wxl** [comment]
@tsujan understood though I've certainly waited the thousands of years it takes for Debian to install just to check an LXQt bug before reporting it 😂 Sometimes starting from nothing is the only way to be sure. And sometimes you can't reproduce it from scratch, in which case you know it's some quirk of your personal system which, while harder to unravel, certainly points away from the target of the issue at hand…

**Comment by tsujan** [comment]
@wxl Actually, I took your advice. Started to install Tumbleweed on VirtualBox. The installer says, "Remaining: 3.964 GiB..." ;)

**Comment by tsujan** [comment]
rmp and deb based distros have very slow package managers. The fastest I've seen is pacman (belonging to Arch); it's ~7 times faster than apt.

**Comment by trmdi** [comment]
Thanks everyone for your confirmation.

@NickLion: I don't know from when this bug has happened, according to @tsujan it is a bug in Qt 5.13.

@tsujan: zypper is fast too. I'm not sure if it is as fast as pacman though.

Btw, you can download the Tumbleweed KDE Live iso which is ~900MB to test. If you `dd` it to a USB it will run in the persistent mode, but I don't know how to dd it to a virtual disk image for VirtualBox.

**Comment by tsujan** [comment]
@trmdi Unfortunately, the installer has had a hang with `yast2-metapackage-handler`. I should abort after 30 minutes :(

**Comment by wxl** [comment]
@tsujan I'd say the onus to show the reproducibility lies in the reporter!

@trmdi you don't need to do anything. Just stick the ISO in the virtual optical drive.

As always, make sure to check your hashes before doing anything…

**Comment by tsujan** [comment]
@wxl I know this for sure: if the installer hangs for the second time, I'll press `Shift+Delete`.

**Comment by trmdi** [comment]
@wxl I meant to use the live image in the persistent mode. I don't think it will save your changes if you mount it into the virtual optical drive.

**Comment by tsujan** [comment]
@trmdi Do you know why `yast2-metapackage-handler` hangs? It happened again! I installed Tumbleweed on VirtualBox last year without any problem (sadly, I removed its file).

**Comment by trmdi** [comment]
@tsujan I don't know, I think you can try the Live KDE iso for another quick test. :D

**Comment by tsujan** [comment]
Unbelievable! After 15 minutes, `yast2-metapackage-handler` was installed. Remaining: 1.199 GiB! I love `pacman` ;)

@trmdi What I'm trying to do isn't possible with a live ISO.

**Comment by trmdi** [comment]
Ok, hope you could reproduce it.

**Comment by tsujan** [comment]
You're very optimistic. I'll be glad if the installation finishes and I can start Tumbleweed ;) Remaining: 753.13 MiB...

**Comment by tsujan** [comment]
OK, let me share my looong experience with you:

At last, I installed Tumbleweed on VirtualBox — I may have been spoiled by Arch/Manjaro but, frankly, it was a torture.

Then I started it. It took 5 minutes to reach the login screen and 4 more minutes to log in.

Then I installed VitualBox additions on it and rebooted. Success :)

Then I installed some dependencies (why did zypper install the whole Qt doc?!), compiled and installed the latest git Kvantum manually, changed the theme to KvMojave, started Falkon and… see what I saw with compositing enabled (Dolphin proves that):

![falkon](https://user-images.githubusercontent.com/981076/66281907-3b73e680-e8ca-11e9-8eb9-3d26dc3d5c69.png)

In short: there was no problem at all.

I close this but feel free to add comments or try to prove that I'm wrong (in which case, I'll reopen it).

Side note: I don't understand why OpenSUSE's installer doesn't let the user choose what he wants. I remember Fedora Core 4 with 5 CDs (14 years ago?), whose installer let me choose almost every app and, although I was just a newbie, the installation went smooth.

**Comment by tsujan** [comment]
Ah, I found the difference: I have Preferences → General → After launch → Open blank page. You have … → Restore session. Now, I can reproduce it with "Restore session". Reopening...

**Comment by trmdi** [comment]
Nice. But isn't "Restore session" the default option of Falkon? You changed it before the test? And how do you know that option relates to the issue? Just curious.

**Comment by tsujan** [comment]
> But isn't "Restore session" the default option of Falkon? You changed it before the test?

Yes, it is and no, I didn't change it under OpenSUSE but there's nothing to restore on the first run ;) The above screenshot was taken when I started Falkon for the first time. Of course, I changed the setting under my distro, Manjaro, years ago; that was why I couldn't reproduce the problem.

I know exactly where the problem is: they changed `QWidget::setAttribute` in Qt 5.13.1 and caused it. I can simply generalize the workaround that I've already included in Kvantum and get rid of the issue but I'm searching for a deeper workaround.

I'll tell you as soon as I make a decision.

**Comment by tsujan** [comment]
@trmdi 

I went with "the deeper workaround". It works here but please test it by installing the latest git Kvantum — and please update your rpm's and close this page if your test is successful.

It also solves another issue that I encountered after upgrading to Qt 5.13.1. To me, 5.13 was the worst version of Qt5 (it had [a third bug](https://bugreports.qt.io/browse/QTBUG-77115) that I reported and they fixed in 5.13.2).

**Comment by trmdi** [comment]
Works well now. Thank you! 👍

**Comment by tsujan** [comment]
Thanks for reporting and testing!

**Comment by tsujan** [comment]
Mainly because of this and another workaround, I published 0.12.0 right now.

**Comment by trmdi** [comment]
// Off-topic:
Sorry to bother you, I think you're very good at Qt so I ask here.
Do you have any way to fix this bug: https://bugreports.qt.io/browse/QTBUG-82934 ?
It's trivial but I really want it to be fixed. :D

**Comment by tsujan** [comment]
@trmdi, I like to work around Qt bugs in Kvantum *if possible* but, frankly, I didn't understand your report at Qt bug tracker (and I doubt that Qt devs understand it). Could you please describe the Qt bug in a step-by-step way so that I can see it here?

**Comment by trmdi** [comment]
Steps to reproduce: drag a file, hold the mouse button and do not move it, then hold the Ctrl key, the cursor will not change until you move the mouse at least 1 px.
It should change without having to move the mouse.

**Comment by tsujan** [comment]
Now I'm working under LXQt and the cursor doesn't change after moving the mouse either -- neither in Dolphin nor in pcmanfm-qt. So, I think that's platform-specific.

But a question (maybe, a dumb question from your standpoint): why should the cursor be changed at all?

**Comment by trmdi** [comment]
Not sure about `pcmanfm-qt`. Bu if you use Dolphin , for example, you drag a file to a directory, then you suddenly realize that you want to move it so you press and hold Shift before release the mouse buton to drop it (without moving the mouse any px). The dropAction will be MoveAction, but the cursor is still the cursor of the CopyAction.

**Comment by tsujan** [comment]
Oh, then it's about a code in Dolphin and so, Dolphin is responsible for changing the cursor -- it isn't automatic and, as you may already know, we haven't implemented this behavior in pcmanfm-qt.

Now, if Dolphin really changes the cursor but the cursor remains the same, then we could talk about a bug in Qt.

But, with my cursor set (Comix) and under LXQt, I don't see a cursor change in Dolphin even after moving the cursor. Therefore, *either* there's something in KDE that does so *or* Comix lacks a kind of cursor.

All in all, the problem is more complex than it may seem at first. Have reported the issue to KDE? If you have but they say it's a Qt bug, the problem gets even more confusing ;)

**Comment by trmdi** [comment]
Yes, I've even sending a patch for this: https://phabricator.kde.org/D28017
Change the cursor while moving the mouse is easy, because QSimpleDrag update it when there is a MouseMove event (as I post the link in the bug report to Qt) but without the moving is harder, and I found a way in the patch above but unfortunately it doesn't work in wayland.
So I think it's better to be fixed in Qt's QDrag.

**Comment by tsujan** [comment]
> it doesn't work in wayland.

People don't believe me when I say Wayland is far from being ready for real use: *Wayland still has lots of problems.* You may use your apps in a way that you don't encounter those problems but they exist. If I had time and wanted to tell you about the serious problems of Plasma Wayland, I should write a very long comment.

**Comment by tsujan** [comment]
And QDrag is one of those problems. That's why I've disabled FeatherPad's tab DND under Wayland.

**Comment by tsujan** [comment]
@trmdi 
Wayland aside, to my surprise, I saw Shift/Ctrl+Drag in pcmanfm-qt today. I'd never used Shift/Ctrl+Drag before and I think many users aren't aware of it either. Strangely, I didn't see it yesterday -- maybe because I was busy with coding.

We haven't added any code for Shift/Ctrl+Drag to `libfm-qt`. So, contrary to what I thought, this is the default behavior of (some) Qt views, which our subclassed view has inherited.

As for the bug you found, it happens with pcmanfm-qt too. I think it can be fixed in the class somehow. `libfm-qt` subclasses `QListView` and `QTreeView`. I'll investigate if a simple workaround is possible in `libfm-qt`. I should also add an explanation for users....

Thanks for the valuable info!
